### PR TITLE
Encode `/quote` response JIT orders and pre-interactions

### PR DESCRIPTION
**Pull request by squadgazzz**
# Description
Follow up to #3033, which uses the newly added `/quote` response fields with pre-interaction and JIT orders, encodes them into the settlement in order to verify the provided quote with this data.

## How to test
New e2e tests:
1. Ensures that an order can be fulfilled using the CoW AMM liquidity only.
2. Tests the quote encoding logic.

### CODE REVIEW

File: crates/e2e/tests/e2e/cow_amm.rs [hunk 1]
```diff
@@ -595,3 +596,685 @@ async fn cow_amm_driver_support(web3: Web3) {
...
+    // Fund trader "bob" with DAI and approve allowance
+    dai.mint(bob.address(), to_wei(250)).await;
+    tx!(
+        bob.account(),
+        dai.approve(onchain.contracts().allowance, U256::MAX)
+    );
+
+    // todo: figure out why this is even required.
+    dai.mint(onchain.contracts().gp_settlement.address(), to_wei(30))
+        .await;
```
Comment (squadgazzz) [inline crates/e2e/tests/e2e/cow_amm.rs:1209]: The PR shouldn't be merged until this is resolved.

Comment (squadgazzz) [reply crates/e2e/tests/e2e/cow_amm.rs:1209]: Can't reproduce the revert when the settlement contract is not funded enough using a manual rpc call with either anvil or forked tenderly node. So weird. Tried different options with/without state overrides, different block numbers. All the same: the simulation passes, but not in the test...

File: crates/e2e/tests/e2e/cow_amm.rs [hunk 2]
```diff
@@ -595,3 +596,685 @@ async fn cow_amm_driver_support(web3: Web3) {
...
+                },
+                executed_amount: cow_amm_order.sell_amount - fee_cow_amm,
+                fee: Some(fee_cow_amm),
+            }),
+            // todo: this should be removed by providing a surplus capturing owners list(https://github.com/cowprotocol/services/pull/3048)
+            solvers_dto::solution::Trade::Fulfillment(solvers_dto::solution::Fulfillment {
+                order: [0u8; 56],
+                executed_amount: to_wei(230) - fee_user,
+                fee: None,
+            }),
```
Comment (squadgazzz) [inline crates/e2e/tests/e2e/cow_amm.rs:1250]: This dummy trade is required to pass the following check in order to accept the solution:
https://github.com/cowprotocol/services/blob/e1c93e92f2aeb9bcfda4d758242f8984d9004bed/crates/driver/src/domain/quote.rs#L122-L131

All the regular trades are still skipped as it was before when building a quote response:
https://github.com/cowprotocol/services/blob/e1c93e92f2aeb9bcfda4d758242f8984d9004bed/crates/driver/src/domain/quote.rs#L39-L86

File: crates/e2e/tests/e2e/cow_amm.rs [hunk 3]
```diff
@@ -595,3 +596,685 @@ async fn cow_amm_driver_support(web3: Web3) {
...
+        .unwrap();
+    let cow_amm = contracts::CowAmm::at(&web3, cow_amm_address);
+
+    // Start system with the regular baseline solver as a quoter but a mock solver
+    // for the actual solver competition. This allows us to handcraft a solution.
+    let mock_solver = Mock::default();
+    colocation::start_driver(
+        onchain.contracts(),
+        vec![
+            colocation::start_baseline_solver(
```
Comment (MartinquaXD) [inline crates/e2e/tests/e2e/cow_amm.rs:718]: We are not using the regular baseline solver for anything anymore, right? Then this can be removed and the comment above be amended.

File: crates/e2e/tests/e2e/cow_amm.rs [hunk 4]
```diff
@@ -595,3 +596,685 @@ async fn cow_amm_driver_support(web3: Web3) {
...
+        valid_to: model::time::now_in_epoch_seconds() + 300,
+        kind: OrderKind::Sell,
+        ..Default::default()
+    }
+    .sign(
+        EcdsaSigningScheme::Eip712,
+        &onchain.contracts().domain_separator,
+        SecretKeyRef::from(&SecretKey::from_slice(bob.private_key()).unwrap()),
+    );
+    let user_order_id = services.create_order(&user_order).await.unwrap();
```
Comment (MartinquaXD) [inline crates/e2e/tests/e2e/cow_amm.rs:877]: To check that the quote actually gets verified it would be good to first get a quote from the API and assert that it has `verified: true`. Then you can call `create_order()` and pass the generated quote id.
Should effectively not change anything but it would make it clear in the test that checking that the quote was verified is important.

File: crates/e2e/tests/e2e/cow_amm.rs [hunk 5]
```diff
@@ -595,3 +596,685 @@ async fn cow_amm_driver_support(web3: Web3) {
...
+            .balance_of(bob.address())
+            .call()
+            .await
+            .unwrap();
+
+        let amm_weth_sent = amm_weth_balance_before - amm_weth_balance_after;
+        let bob_weth_received = bob_weth_balance_after - bob_weth_balance_before;
+
+        // Bob should receive WETH, CoW AMM's WETH balance decreases
+        bob_weth_received >= user_order.buy_amount && amm_weth_sent >= cow_amm_order.sell_amount
```
Comment (MartinquaXD) [inline crates/e2e/tests/e2e/cow_amm.rs:944]: If you configure a sell order you should never have more tokens leave your account than you signed.
```suggestion
        bob_weth_received >= user_order.buy_amount && amm_weth_sent == cow_amm_order.sell_amount
```

File: crates/e2e/tests/e2e/cow_amm.rs [hunk 6]
```diff
@@ -595,3 +596,685 @@ async fn cow_amm_driver_support(web3: Web3) {
...
+}
+
+#[tokio::test]
+#[ignore]
+async fn local_node_cow_amm_quoting() {
+    run_test(cow_amm_quoting).await;
+}
+
+/// Tests quotes verification that contains CoW AMM JIT orders.
+async fn cow_amm_quoting(web3: Web3) {
```
Comment (MartinquaXD) [inline crates/e2e/tests/e2e/cow_amm.rs:966]: I see there is a separate test for quoting specifically now. Can this not be merged with the other new test?

File: crates/e2e/tests/e2e/cow_amm.rs [hunk 7]
```diff
@@ -595,3 +596,685 @@ async fn cow_amm_driver_support(web3: Web3) {
...
+            .approve(cow_amm_factory.address(), to_wei(1))
+    );
+
+    tx_value!(
+        solver.account(),
+        to_wei(1),
+        onchain.contracts().weth.deposit()
+    );
+
+    // Fund the settlement contract with WETH so it can pay out the user order.
```
Comment (MartinquaXD) [inline crates/e2e/tests/e2e/cow_amm.rs:1027]: Should also not be needed AFAIU.

File: crates/e2e/tests/e2e/cow_amm.rs [hunk 8]
```diff
@@ -595,3 +596,685 @@ async fn cow_amm_driver_support(web3: Web3) {
...
+        .unwrap();
+    let cow_amm = contracts::CowAmm::at(&web3, cow_amm_address);
+
+    // Start system with the regular baseline solver as a quoter but a mock solver
+    // for the actual solver competition. This allows us to handcraft a solution.
+    let mock_solver = Mock::default();
+    colocation::start_driver(
+        onchain.contracts(),
+        vec![
+            colocation::start_baseline_solver(
```
Comment (MartinquaXD) [inline crates/e2e/tests/e2e/cow_amm.rs:1084]: No `test_solver` used here either AFAICS.

File: crates/shared/src/price_estimation/trade_verifier.rs [hunk 9]
```diff
@@ -400,16 +414,86 @@ fn encode_settlement(
...
+            sell_amount: jit_order.sell_amount,
+            buy_amount: jit_order.buy_amount,
+            valid_to: jit_order.valid_to,
+            app_data: AppDataHash::from_str(&jit_order.app_data)?,
+            fee_amount: 0.into(),
+            kind: match &jit_order.side {
+                dto::Side::Buy => OrderKind::Buy,
+                dto::Side::Sell => OrderKind::Sell,
+            },
+            partially_fillable: false,
```
Comment (MartinquaXD) [inline crates/shared/src/price_estimation/trade_verifier.rs:433]: This is a slightly problematic detail I didn't consider before. The JIT order API for the `/settle` call doesn't allow for partially fillable orders because that assumes only truly new JIT orders being proposed there and for those fill-or-kill should be sufficient.
But for the `/quote` endpoint the JIT orders returned could also be user orders (masked as JIT orders to not introduce even more enums). So if a solver would like to return a quote based on a partially fillable user order in the book this would not work here.
I think the API needs to allow sending the `partially_fillable` flag with the JIT orders.

File: crates/shared/src/price_estimation/trade_verifier.rs [hunk 10]
```diff
@@ -400,16 +414,86 @@ fn encode_settlement(
...
+                let signature =
+                    Signature::from_bytes(jit_order.signing_scheme, Vec::new().as_slice())?;
+                (owner, signature)
+            }
+            _ => {
+                let signature =
+                    Signature::from_bytes(jit_order.signing_scheme, &jit_order.signature)?;
+                let owner = signature
+                    .recover(domain_separator, &order_data.hash_struct())?
+                    .unwrap()
```
Comment (MartinquaXD) [inline crates/shared/src/price_estimation/trade_verifier.rs:455]: Since we already have error handling in the function we might as well replace this `.unwrap()` with a `.context(<ERROR_MESSAGE>)?`.

File: crates/shared/src/price_estimation/trade_verifier.rs [hunk 11]
```diff
@@ -400,16 +414,86 @@ fn encode_settlement(
...
+            }
+        };
+
+        tokens.push(jit_order.sell_token);
+        tokens.push(jit_order.buy_token);
+
+        match jit_order.side {
+            dto::Side::Sell => {
+                clearing_prices.push(jit_order.buy_amount);
+                clearing_prices.push(jit_order.executed_amount);
```
Comment (MartinquaXD) [inline crates/shared/src/price_estimation/trade_verifier.rs:467]: @harisang in this PR we are testing that quotes can be settled (and verified) against JIT orders. We only introduced the concept of JIT orders in the `/quote` response because solvers could also settled against user orders by indexing the orderbook and pretending the user order is a JIT order.
In this particular part this would cause JIT orders to be settled at the limit price in the settlement. So if a solver returns a regular user order as a JIT order which would not reflect what would happen in a real settlement entirely correctly. I assumed this should be fine but I actually never asked the solver team about it.
Do you foresee any significant issues from this detail? Asking because we'd otherwise have to be able to know which JIT orders are actually surplus capturing in the quote verification which would be a huge PITA.

File: crates/shared/src/trade_finding/external.rs [hunk 12]
```diff
@@ -232,8 +236,7 @@ mod dto {
         pub executed_amount: U256,
         pub receiver: H160,
         pub valid_to: u32,
-        #[serde_as(as = "BytesHex")]
-        pub app_data: Vec<u8>,
```
Comment (MartinquaXD) [inline crates/shared/src/trade_finding/external.rs:236]: What's the reason for this change? The original version is more correct, no?

File: crates/e2e/tests/e2e/cow_amm.rs (hunk 2 above)
Comment (MartinquaXD) [reply crates/e2e/tests/e2e/cow_amm.rs:1250]: Is this actually a dummy trade? Isn't this how a real solver would actually indicate that it wants to settle the fake order generated for the quote?

Comment (squadgazzz) [reply crates/e2e/tests/e2e/cow_amm.rs:1250]: I don't really see this trade used in the quote response at all.

File: crates/e2e/tests/e2e/cow_amm.rs (hunk 3 above)
Comment (squadgazzz) [reply crates/e2e/tests/e2e/cow_amm.rs:718]: It is used for native price fetching.

File: crates/e2e/tests/e2e/cow_amm.rs (hunk 4 above)
Comment (squadgazzz) [reply crates/e2e/tests/e2e/cow_amm.rs:877]: I forgot to mention that in the description. I wanted to merge these two e2e tests into a single one, but that also requires changes in the mocked solver. The PR is already huge, maybe it makes sense to accomplish this as a follow-up.

File: crates/e2e/tests/e2e/cow_amm.rs (hunk 6 above)
Comment (squadgazzz) [reply crates/e2e/tests/e2e/cow_amm.rs:966]: https://github.com/cowprotocol/services/pull/3038#discussion_r1796780130

File: crates/shared/src/price_estimation/trade_verifier.rs (hunk 11 above)
Comment (squadgazzz) [reply crates/shared/src/price_estimation/trade_verifier.rs:467]: In the competition encoding, executed amount is not used here:
https://github.com/cowprotocol/services/blob/559ebf3dce53837f660de743bbb55523e4e3a1c9/crates/driver/src/domain/competition/solution/encoding.rs#L116-L125

Maybe that is why the e2e test requires settlement contract to be funded with DAI:
https://github.com/cowprotocol/services/blob/8fd8cf60c823a024f9d6718634ece600124b17b4/crates/e2e/tests/e2e/cow_amm.rs#L1208-L1210

But when using plain sell/buy amounts, the simulation gets reverted with a vague error, so I couldn't understand what exactly is wrong with that case.

File: crates/shared/src/trade_finding/external.rs (hunk 12 above)
Comment (squadgazzz) [reply crates/shared/src/trade_finding/external.rs:236]: The reason is that the AppDataHash needs to be constructed later here:
https://github.com/cowprotocol/services/blob/0617f4f92d9d2dd9d6c3a160fa239d1fc6824f63/crates/shared/src/price_estimation/trade_verifier.rs#L429
Which means that the Vec<u8> needs to be encoded back to String and only then to the AppDataHash. The change avoids back and forth decoding/encoding of the original String from the response.

File: crates/e2e/tests/e2e/cow_amm.rs [hunk 13]
```diff
@@ -595,3 +596,685 @@ async fn cow_amm_driver_support(web3: Web3) {
...
+            .approve(cow_amm_factory.address(), to_wei(1))
+    );
+
+    tx_value!(
+        solver.account(),
+        to_wei(1),
+        onchain.contracts().weth.deposit()
+    );
+
+    // Fund the settlement contract with WETH so it can pay out the user order.
```
[the remark this answers is not in the document]
Comment (squadgazzz) [reply crates/e2e/tests/e2e/cow_amm.rs:661]: Then, I assume there is still something that needs to be corrected since the simulation fails without this funding for both quote and settlement e2e tests.

File: crates/e2e/tests/e2e/cow_amm.rs (hunk 3 above)
Comment (squadgazzz) [reply crates/e2e/tests/e2e/cow_amm.rs:718]: https://github.com/cowprotocol/services/blob/af9d499618e4e01b22952cd5d240e42b93d9442e/crates/e2e/src/setup/services.rs#L107-L109

File: crates/shared/src/price_estimation/trade_verifier.rs (hunk 9 above)
Comment (squadgazzz) [reply crates/shared/src/price_estimation/trade_verifier.rs:433]: That means, the solvers API needs to be also changed since now it doesn't contain this field in the JitOrder json:
https://github.com/cowprotocol/services/blob/39b3f5bda1d3909d0e07aa0f0fe3f8a460a014f2/crates/driver/src/infra/solver/dto/solution.rs#L278-L298

File: crates/e2e/tests/e2e/cow_amm.rs [hunk 14]
```diff
@@ -595,3 +596,685 @@ async fn cow_amm_driver_support(web3: Web3) {
...
+        dai.approve(onchain.contracts().allowance, U256::MAX)
+    );
+
+    // todo: figure out why this is even required.
+    dai.mint(onchain.contracts().gp_settlement.address(), to_wei(30))
+        .await;
+
+    // Compensate a delay between the `CurrentBlockStream` and the actual onchain
+    // data.
+    tokio::time::sleep(tokio::time::Duration::from_secs(2)).await;
```
[the remark this answers is not in the document]
Comment (squadgazzz) [reply crates/e2e/tests/e2e/cow_amm.rs:1213]: Autopilot receives onchain data using `CurrentBlockStream,` which has a delay between the actual onchain operations making the test fail since autopilot doesn't get informed about all the operations above.

File: crates/driver/src/domain/competition/solution/encoding.rs [hunk 15]
```diff
@@ -312,7 +312,7 @@ struct Flags {
     buy_token_balance: order::BuyTokenBalance,
 }
 
-mod codec {
+pub mod codec {
```
[the remark this answers is not in the document]
Comment (squadgazzz) [reply crates/driver/src/domain/competition/solution/encoding.rs:315]: `super` is required since it is located in the `/driver/competition/..`, while I need to used from the `/driver/domain/..`.

File: crates/shared/src/price_estimation/trade_verifier.rs (hunk 9 above)
Comment (MartinquaXD) [reply crates/shared/src/price_estimation/trade_verifier.rs:433]: It was intended like this in the `/solve` API but I guess it makes sense to keep these 2 APIs the same.

File: crates/e2e/tests/e2e/cow_amm.rs (hunk 1 above)
Comment (squadgazzz) [reply crates/e2e/tests/e2e/cow_amm.rs]: [`17292a6` (#3038)](https://github.com/cowprotocol/services/pull/3038/commits/17292a6296958143a451675566026b93e5168141) fixes this 🤷‍♂️

File: crates/e2e/tests/e2e/cow_amm.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (MartinquaXD) [reply crates/e2e/tests/e2e/cow_amm.rs:1211]: This data only exists in memory. BTW we could also try to reduce the `--block-stream-poll-interval`. It's currently 1s which seems unnecessarily high.

File: crates/e2e/tests/e2e/cow_amm.rs (hunk 2 above)
Comment (MartinquaXD) [reply crates/e2e/tests/e2e/cow_amm.rs:1248]: The relevant part is [here](https://github.com/cowprotocol/services/blob/main/crates/driver/src/domain/quote.rs#L137). There the driver turns a `/quote` requests into a fake order with `uid: 0x000...000`. If the solver is able to provide a quote it will return a solution that settles order `0x000...000`.

Comment (squadgazzz) [reply crates/e2e/tests/e2e/cow_amm.rs]: Yes, so the initially mentioned "dummy" trade remains unused. Should it now be used instead of the fake order?

File: crates/e2e/tests/e2e/cow_amm.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (squadgazzz) [reply crates/e2e/tests/e2e/cow_amm.rs]: Reduced both the config value and the delay.

File: crates/shared/src/trade_finding/external.rs (hunk 12 above)
Comment (squadgazzz) [reply crates/shared/src/trade_finding/external.rs:236]: True. How about using the `AppDataHash` directly since it already implements a `Deserializer`:
[`9d23043` (#3038)](https://github.com/cowprotocol/services/pull/3038/commits/9d23043cddc515c964d68f77f718cce0fa8996d2)

**Comment by squadgazzz** [comment]
Will be reimplemented in scope of the https://github.com/cowprotocol/services/issues/3082
